### PR TITLE
core: prevent USB events from waking autolock timer

### LINKED ISSUE #2651: Softlock behaves incorrectly with CoinJoin
from Slack:

> Apparently the soft lock during coinjoin behaves somewhat unexpectedly. I left the device alone with lock time 10 min (afaik default) and after few hours, the device is still not locked. Looks like the API calls around coinjoin are resetting internal counter, which they shouldn't IMO.

Presumably the device should not be usable, but CoinJoin messages should be allowed to run through.

With current softlock implementation any USB message will reset the timer.
We might want to differentiate "soft" lock touch (done by an USB message) from a "hard" touch (done by user interaction).
A "hard" touch resets the internal timer. A "soft" touch keeps the timer running, but if there are any "soft" touches registered at the end of the period, the locking will not shut down the current workflow.

To test: CoinJoin operations (via `DoPreauthorized`) are allowed to run even if the device is currently locked. Does this actually work, i.e., doesn't the signing process somehow implicitly rely on storage being unlocked?

### PR DESCRIPTION
**Pull request by mmilata**
~~Given the timeline and complexity I did not feel confident implementing the proposed solution of #2651. This PR introduces different solution which avoids touching the idle timer when processing message listed as allowed to be processed when device is locked.~~

~~The disadvantage is that with unfortunate timing the autolock interrupts coinjoin signing if there's one in progress.~~

Fixes #2651.

### CODE REVIEW

**Comment by matejcik** [comment]
i'm not sure if that's better or worse than the current state.

what if, instead, we just force locked when going to the coinjoin busy state?
what do you think @andrewkozlik ?

**Comment by andrewkozlik** [comment]
> i'm not sure if that's better or worse than the current state.

I think worse.

> what if, instead, we just force locked when going to the coinjoin busy state? what do you think @andrewkozlik ?

I think this is more of a product / UX design decision, but I personally am not in favor of the idea.

**Comment by matejcik** [comment]
we're converging on the following:

1. keep the modification that preserves deadline across session restarts
2. remove "touch idle timer on usb message"
3. allow `DoPreauthorized` to set a flag, which, when set, causes the locking function to _not_ shutdown the running workflow (but otherwise lock the config, update the homescreen for next run, etc.)

File: core/src/trezor/workflow.py
Comment (matejcik) [inline core/src/trezor/workflow.py:202]: this seems wrong? means that in the weird case where a dont_autolock workflow is waiting for user confirmation, it will remain unlocked.

Comment (mmilata) [reply core/src/trezor/workflow.py:202]: Isn't that what we want though, for the whitelisted workflows, to remain unlocked? With locked storage.

I've used the the same whitelist for not waking up the timer because I'm not sure whether we can entirely ignore usb messages. That would break cases like `test_lots_of_inputs` with short autolock delay, which currently work fine?

Comment (mmilata) [reply core/src/trezor/workflow.py]: Ok, let's handle it in apps/bitcoin ...

Comment (matejcik) [reply core/src/trezor/workflow.py]: > That would break cases like test_lots_of_inputs with short autolock delay, which currently work fine?

right, which is why I wanted bitcoin signing to set its own exception to autolock too.

ignoring USB messages seems overall more correct, if we have a mechanism to let the workflow finish

File: core/src/trezor/wire/__init__.py
Comment (matejcik) [inline core/src/trezor/wire/__init__.py:379]: this might be better in `workflow.start_default()` ?

**Comment by mmilata** [comment]
Tests failing ... looks like preserving deadline across restarts is somehow broken. Did some brief testing on device and didn't catch this.

**Comment by mmilata** [comment]
When user changed the autolock delay the value didn't propagate to the timer correctly as it used the cached value ... db44cb6aad4b7505f1f95e2c373b411c09cefd67

**Review by matejcik — APPROVED** [review]
preliminary ACK.

i would like to see some tests, let's talk tomorrow whether we can quickly add some or whether we'll defer that to a sepaeate issue.

File: core/src/apps/base.py
Comment (andrewkozlik) [inline core/src/apps/base.py:304]: So this means that a `LockDevice` message will interrupt every workflow. Do we want that? I think yes, but just checking.

Comment (andrewkozlik) [reply core/src/apps/base.py:304]: Actually sending any unexpected message in a workflow will probably interrupt it IIRC, won't it? So the above though is probably irrelevant. I was just trying to understand how this parameter is used. Anyway, it's OK as is.

File: core/src/apps/bitcoin/sign_tx/bitcoin.py
Comment (andrewkozlik) [inline core/src/apps/bitcoin/sign_tx/bitcoin.py:83]: So now if a user leaves the Trezor hanging in an output confirmation screen, leaves the room and an attacker comes in, the attacker can finish the transaction confirmation even if the device has been autolocked. In comparison if the user leaves the Trezor hanging in a change label screen, i.e. a far less security-sensitive operation, the device will stop the workflow after the autolock timeout and won't let a potential attacker finish the label change.

**Comment by matejcik** [comment]
pushed a bunch of tests confirming the new behavior in 97b6bde.

still not sure how to do the DoPreauthorized one -- maybe i'll need to extend one of the existing coinjoin tests

File: core/src/apps/bitcoin/sign_tx/bitcoin.py
Comment (andrewkozlik) [reply core/src/apps/bitcoin/sign_tx/bitcoin.py:83]: Oh yeah, you are absolutely right. As usual I am doing two things at once and I overlooked that steps 1, 2 & approval are already complete at this stage.

> only if a prompt pops up afterwards, then it would remain open. (something something late path warning? or did we already resolve that?)

That would be a problem, but I think that's already resolved. All checks and warnings should pop up in Steps 1, 2 and the approval step. After that if some check fails, then the entire signing process should fail.

Comment (andrewkozlik) [reply core/src/apps/bitcoin/sign_tx/bitcoin.py:83]: For example `BasicApprover.add_internal_input()` shows a warning but `BasicApprover.check_internal_input()` raises an exception. So I think we are fine.

File: core/src/storage/cache.py
Comment (mmilata) [reply core/src/storage/cache.py]: [fba8e56](https://github.com/trezor/trezor-firmware/pull/2933/commits/fba8e5650137ff11d07e5a96270fa05add22dfea) changes this to
> Persist autolock deadline across main loop restart between workflows.

Hopefully that's better. We don't have anything about workflow restarts on docs.trezor.io, right?

File: core/src/trezor/workflow.py
Comment (andrewkozlik) [inline core/src/trezor/workflow.py:205]: This docstring probably needs an update. USB messages no longer count as activity.

Comment (andrewkozlik) [reply core/src/trezor/workflow.py:205]: Also please document `restore_deadline`.

File: core/src/trezor/workflow.py
Comment (andrewkozlik) [inline core/src/trezor/workflow.py:237]: Again, the docstring above `idle_timer.set() also counts as an activity, so all running idle timers are reset.` needs updating if I understand it correctly.

Comment (andrewkozlik) [reply core/src/trezor/workflow.py:237]: Also the comment below the docstring `The reason for counting set() [...]` is probably invalidated by these changes.

File: core/src/trezor/workflow.py
Comment (andrewkozlik) [reply core/src/trezor/workflow.py:205]: Still trying to understand this. In practice there is only one task in `self.tasks`, the `lock_device_if_unlocked()`, right? But the function is written like there could be more tasks, each one with it's own timeout. Now we are somehow mixing up all the deadlines together with `storage.cache.autolock_deadline` even though they can be different?

**Comment by mmilata** [comment]
6af361aeed156489bd654dbcfeaef746869d39e1 adds DoAuthorized click test

**Review by matejcik — APPROVED** [review]
LGTM now

**Pull request merged**
